dieboljo/embedded-loop-station
# Feature/sync display audio

## Related Issues:
* [Multi-track controller](https://app.asana.com/0/1203221076104575/1204059708717193)
* [Adjust save loop function](https://app.asana.com/0/1203221076104575/1204475433853819)
* [Adjust track gain](https://app.asana.com/0/1203221076104575/1204445914333577)
* [Dampen output when recording](https://app.asana.com/0/1203221076104575/1204475433853814)

## Main Changes:
* Moved track functions under the umbrella of a controller class. Primarily, this class acts a middleman that routes actions to the proper track and synchronizes the tracks. Four tracks are functional, each capable of stereo input.
* Refactored some methods of the Display class to separate state updates from drawing. This is necessary because the Display class is used for both input and output, and splitting their duties allows other parts of the app to "link in".
* Added the ability to load loops into a single track.
* Modified the pan functionality to be active even when a track is not recording. This matches the behavior of most audio tools. While the ability to have differently panned audio in a single track was nice, with multiple tracks it isn't necessary and likely would have confused the user.
* Added functionality that dampens perceived output for the selected track when recording in overdub mode, and silences it in replace mode. This only affects the selected track, and it doesn't affect the recorded audio.
* Added the ability to change a single track's output gain. This affects perceived output and recorded output.
* Updated the displayed track name to only show the current track, rather than the underlying file. Also, the location in the display that shows the track name is no longer used to display other statuses. If we want to show toast messages, we may need to designate a different location.

## Steps To Test:
1. Record a loop, then save it to a file.
2. Record over the selected track.
3. Load the saved loop into the same track, and confirm it loads correctly.
4. Test recording on different tracks, and verify that recorded audio doesn't tracks other than the one selected.
5. Test the new implementation of the pan effect.
6. Test the dampening effect when recording, and confirm that the dampening doesn't affect the recorded audio.

## TODO:
- [ ] I wasn't able to get the `resize` or `addValue` functions to work with the library, so for now the array is reinitialized when a loop is saved. This way, the program doesn't need to be restarted for the loop to show in the library. Ideally, I think the `fileArray` should start with a size of 4, and can be added to at will. The program will probably crash if there are more than 20 loops in the `/loops` directory.
- [ ] It seems that the _fade_ and _pan_ knobs are more sensitive at the lower ends of their range than their upper ends. This is the case even when using the `map` function. We could try to figure out a calculation that addresses this.
